File: app/src/main/java/net/feheren_fekete/applist/applistpage/model/SectionData.java
package net.feheren_fekete.applist.applistpage.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class SectionData extends BaseData {
    private String mName;
    private List<StartableData> mStartables;
    private boolean mIsRemovable;
    private boolean mIsCollapsed;

    public SectionData(long id,
                       String name,
                       List<StartableData> startables,
                       boolean isRemovable,
                       boolean isCollapsed) {
        super(id);
        mName = name;
        mStartables = startables;
        mIsRemovable = isRemovable;
        mIsCollapsed = isCollapsed;
    }

    public String getName() {
        return mName;
    }

    public void setName(String name) {
        mName = name;
    }

    public boolean isRemovable() {
        return mIsRemovable;
    }

    public void setCollapsed(boolean collapsed) {
        mIsCollapsed = collapsed;
    }

    public boolean isCollapsed() {
        return mIsCollapsed;
    }

    public boolean isEmpty() {
        return mStartables.isEmpty();
    }

    public List<StartableData> getStartables() {
        return mStartables;
    }

    public List<AppData> getApps() {
        ArrayList<AppData> result = new ArrayList<>();
        for (StartableData startableData : mStartables) {
            if (startableData instanceof AppData) {
                result.add((AppData) startableData);
            }
        }
        return result;
    }

    public void setStartables(List<StartableData> startables) {
        mStartables = startables;
    }

    public void addStartables(int index, List<StartableData> startableDatas) {
        mStartables.addAll(index, startableDatas);
    }

    public void addApps(int index, List<AppData> apps) {
        mStartables.addAll(index, apps);
    }

    public void addStartable(StartableData startableData) {
        mStartables.add(startableData);
    }

    public boolean hasStartable(StartableData startableData) {
        return mStartables.contains(startableData);
    }

    public boolean removeStartable(StartableData startableData) {
        return mStartables.remove(startableData);
    }

    public void sortStartablesAlphabetically() {
        Collections.sort(mStartables, new StartableData.NameComparator());
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || !(o instanceof SectionData)) {
            return false;
        }
        SectionData other = (SectionData) o;
        return mName.equals(other.getName());
    }

    public static final class NameComparator implements Comparator<SectionData> {
        @Override
        public int compare(SectionData lhs, SectionData rhs) {
            return lhs.getName().compareTo(rhs.getName());
        }
    }

}

File: app/src/main/java/net/feheren_fekete/applist/applistpage/model/SectionData.kt
package net.feheren_fekete.applist.applistpage.model

class SectionData(id: Long,
                  var name: String,
                  startables: List<StartableData>,
                  val isRemovable: Boolean,
                  var isCollapsed: Boolean) : BaseData(id) {

    private val startables = mutableListOf<StartableData>().apply {
        addAll(startables)
    }

    fun isEmpty() = startables.isEmpty()

    fun getStartables() = startables

    fun setStartables(startableDatas: List<StartableData>) {
        startables.clear()
        startables.addAll(startableDatas)
    }

    fun addStartables(index: Int, startableDatas: List<StartableData>) {
        startables.addAll(index, startableDatas)
    }

    fun addStartable(startableData: StartableData) {
        startables.add(startableData)
    }

    fun getStartable(startableId: Long) =
            startables.find {
                it.id == startableId
            }

    fun hasStartable(startableId: Long) =
            getStartable(startableId) != null

    fun hasStartable(startableData: StartableData) =
            startables.contains(startableData)

    fun removeStartable(startableId: Long): Boolean {
        val startable = startables.find {
            it.id == startableId
        }
        return if (startable != null) {
            startables.remove(startable)
            true
        } else {
            false
        }
    }

    fun sortStartablesAlphabetically() {
        startables.sortWith(StartableData.NameComparator())
    }

    override fun equals(other: Any?): Boolean {
        if (other == null || other !is SectionData) {
            return false
        }
        return name == other.name
    }

    class NameComparator : Comparator<SectionData> {
        override fun compare(lhs: SectionData, rhs: SectionData): Int {
            return lhs.name.compareTo(rhs.name)
        }
    }

}
